### DISCUSSION TITLE
Problem with duration of stimuli display

### DISCUSSION THREAD

**Discussion by prajwalgatti** [jspsych/jsPsych#761] [Q&A] [answered]
Hi Josh and community,

I have a question regarding the timing of stimulus displays.

I wanted to confirm if the duration of stimuli display was the same as I had defined it in the `trial_duration` parameter. I checked the difference in _time_elapsed_ value between the trial and the previous trial's output data. I found that most trials were consistent with +2 or 3ms, which I think is acceptable, given trial_elapsed is not very accurate (as I had read somewhere).

But my problem is one of the trials exceeds its specified `trial_duration` period by a considerable 20-30ms. 

eg. I had defined trial_duration: 200 but the display duration seemed to be 223ms, 233ms, 226ms, and so on, upon checking several times.

I suspect this problem is because I am displaying two Gabor patch images in this trial, by passing their corresponding base64 strings to an <svg> element for displaying as images. 

Please view the code below, I store the two base64 strings of Gabor patches in `stim` array. The increase in time also scales with how many Gabor patches I am displaying.

``` javascript
var stim_disp = {
    type: 'html-keyboard-response',
    stimulus: 
    '<svg xmlns="http://www.w3.org/2000/svg">'  +
    '<circle class="fixation-point"/>' + 
    '<circle class="reward-cue-left"/>'+
    '<circle class="reward-cue-right"/>'+
    '<image class="left-stim" href="'+ stim[0] + '" style="transform:translate('+ (-stim_radius) +'px,'+ (-stim_radius) +'px)"/>' +
    '<image class="right-stim" href="'+ stim[1] + '" style="transform:translate('+ (-stim_radius) +'px,'+ (-stim_radius) +'px)"/>' +
    '</svg>',
    trial_duration: 200,
    choices: jsPsych.NO_KEYS,
    data : {test_part: 'stimulus_and_cue'},
    on_finish: function(data){
        data.disp_duration = data.time_elapsed - jsPsych.data.get().last(2).values()[0].time_elapsed;
    }
};
```
I would like to know what I can do to resolve this?

Thank you for your answers beforehand!

**Comment by vijaymarupudi** [comment, marked as answer]
I would not use time_elapsed for this purpose. time_elapsed contains the time the browser takes to paint and setup the experiment for each trial. You should use data.rt for your timing needs.

Since in your case, you don't get a keyboard response, I can assure you that the stimulus is actually visible for around 200ms. time_elapsed gives you a measure that includes the time taken by your graphic cards to paint the gabor patch to the screen (during which it is not visible).

**Reply by prajwalgatti** [reply]
Hi Vijay,
I did not know that, thanks for the helpful info! So basically, (if I understood you correctly) there would be a few frames where stimulus would not be visible because it is being drawn by the system.

What I wonder now is if I can make the loading of the Gabor patches more efficient for performance? I currently load them through base64 strings in the HTML tags. Any advice?

**Reply by vijaymarupudi** [reply]
That's the best method I'm aware of!

**Reply by prajwalgatti** [reply]
Instead of base64 strings, I loaded local PNG files of the stimulus and it seems to not have the extended durations anymore!

The stimulus is a square image of about 400px length. I need to do some digging to figure out why this performs better. But is it that using base64 images is a good method for only smaller sized images?

**Comment by vijaymarupudi** [comment]
I think that's because your browser caches the image (the actually display buffer) so that it doesn't have to convert the images to RGB values again. The data uri methods only caches the images (non-parsed data). At least this my theory.
